pink_balancer/inverse_dynamics.py

```diff
@@ -0,0 +1,321 @@
+        J_fl = J_f[:3, leg_indices_v]  # (3, 3)
+
+        if not is_null(v):
+            # Compute the contact Jacobian time variation
+            # (3, nv)
+            Jdot = pin.frameJacobianTimeVariation(
+                self.model, self.data, q, v, foot_frame_id, ref_frame
+            )  # (6, nv)
+
+            Jdot_f = Jdot[:3, self.base_indices_v + leg_indices_v]  # (3, 6)
+
+        # Joint torques due to contact forces
+        joint_torques = np.zeros(3)
+
+        if not is_null(v):
+            v = cast(np.ndarray, v)  # mypy hint: v is *always* np.ndarray here
+            joint_torques += Jdot_f.dot(v[self.base_indices_v + leg_indices_v])
+
+        if not is_null(a):
+            a = cast(np.ndarray, a)
+            joint_torques += J_f.dot(a[self.base_indices_v + leg_indices_v])
+
+        # Project the contact forces onto the joints
+        J_fl = J_f[:3, leg_indices_v]  # (3, 3)
+        J_fl_inv = np.linalg.pinv(J_fl)  # (3, 3)
+
+        M_l = self.data.M[leg_indices_v, leg_indices_v]  # (3, 3)
+
+        tau_contact = M_l @ J_fl_inv @ joint_torques
+
+        return tau_contact
+
+    def cycle(self, observation: dict, dt: float) -> dict:
+        """Compute the inverse dynamics of the given leg."""
+        # Populate the q, v, a arrays from the observation
+
+        leg_indices_q = self.get_leg_indices("both", "config")
+        leg_indices_v = self.get_leg_indices("both", "tangent")
+        for joint_name, joint_idx_q, joint_idx_v in\
+            zip(self.joint_names[1:], leg_indices_q, leg_indices_v):
+            q = observation["servo"][joint_name]["position"]
+            v = observation["servo"][joint_name]["velocity"]
+            tau = observation["servo"][joint_name]["torque"]
+
+            # Finite differences to compute the acceleration
+            a = v - self._v[joint_idx_v] if dt > 0 else 0.0
+            a = a / dt
+
+            # Update the values
+            self._q[joint_idx_q] = q
+            self._v[joint_idx_v] = v
+            self._a[joint_idx_v] = a
+
+            # Update the measured torques
+            self.tau_measured[joint_idx_v] = tau
+
+        # Fill in the base joint values
+        self._q[:3] = 0.0  # We don't observe the base position, and torques\
+        # are not affected by it
+
+        # N.B.: Pinocchio and the IMU use different conventions for quaternions
+        (w, x, y, z) = observation["imu"]["orientation"]
+        R_world_imu = pin.Quaternion(np.array([x, y, z, w]))
+        R_world_base = R_world_imu * self.R_imu_base
+        self._q[3:7] = R_world_base.coeffs() # Quaternion representation
+
+        # Fill in the base velocity values
+        self._v[:3] = 0.0  # We don't observe the base velocity
+        # (we could integrate but it would drift)
+
+        # We convert the angular velocity from the IMU frame to the base frame
+        omega_imu = np.array(observation["imu"]["angular_velocity"])
+        omega_base = self.R_imu_base * omega_imu
```

Comment:
With your notations `R_imu_base` is "rotation from base to IMU", right?

If that's correct, then:
```suggestion
        omega_base = self.R_imu_base.<whatever the quaternion inverse is> * omega_imu
```

(The risk of confusion is the reason why I have decided to always write in full `rotation_A_to_B`.)